### DISCUSSION TITLE
How to use Decompose in an SwiftUI project?

### DISCUSSION THREAD

**Discussion by falcon4ever** [badoo/Decompose#204] [Q&A]
Hi,

On Android/Compose it's pretty easy to get started using:
```kotlin

setContent {
  NavigationSampleTheme {
    val component = RootComponent(defaultComponentContext())
    RootUi(component)
  }
}

@Composable
fun RootUi(component: IRoot) {
// ... compose UI stuff here
}
```

For my demo project, I'm trying to implement the SwiftUI version, but I'm struggling to figure out how to get the default component context.

I looked at the **counter demo** as well as the **compose-jb todoapp** example:
https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/ios/ios/ContentView.swift
https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/ios/ios/ComponentHolder.swift

And it seems like all of them are using a helper class `ComponentHolder`. The new component is created via `DefaultComponentContext` and `LifecycleRegistryKt.LifecycleRegistry()`

When I copy the same code into my SwiftUI project, I unfortunately get a bunch of warnings:
<img width="1229" alt="Screen Shot 2021-08-06 at 8 52 46 PM" src="https://user-images.githubusercontent.com/480019/128582826-ff139621-c2c2-412d-b248-8955b64f09ba.png">

I don't seem to have a `DefaultComponentContext` the closest thing I can find is `DecomposeComponentContext` (renamed from SharedDecomposeComponentContext according to xcode)
<img width="456" alt="Screen Shot 2021-08-06 at 9 03 23 PM" src="https://user-images.githubusercontent.com/480019/128583097-0bd07d80-caf2-4697-abda-65dc21eef850.png">

For the lifecycle, I can't find the LifecycleRegistryKt
<img width="308" alt="Screen Shot 2021-08-06 at 9 03 32 PM" src="https://user-images.githubusercontent.com/480019/128583098-f07a9d61-132a-4383-8115-eaa18804e12f.png">

Are there any additional steps I need to do to get hold of the default component context and the lifecycle registry?

I'm using Decompose 0.3.1 and kotlin v1.5.21

```kotlin
// build.gradle.kts (:shared)
import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget

plugins {
    kotlin("multiplatform")
    kotlin("native.cocoapods")
    id("com.android.library")
    id("kotlin-parcelize")
}

version = "1.0"

kotlin {
    android()

    val iosTarget: (String, KotlinNativeTarget.() -> Unit) -> KotlinNativeTarget =
        if (System.getenv("SDK_NAME")?.startsWith("iphoneos") == true)
            ::iosArm64
        else
            ::iosX64

    iosTarget("ios") {}

    cocoapods {
        summary = "Some description for the Shared Module"
        homepage = "Link to the Shared Module homepage"
        ios.deploymentTarget = "14.1"
        frameworkName = "shared"
        podfile = project.file("../iosApp/Podfile")
    }
    
    sourceSets {
        val commonMain by getting {
            dependencies {
                implementation("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
            }
        }
        val commonTest by getting {
            dependencies {
                implementation(kotlin("test-common"))
                implementation(kotlin("test-annotations-common"))
            }
        }
        val androidMain by getting
        val androidTest by getting {
            dependencies {
                implementation(kotlin("test-junit"))
                implementation("junit:junit:4.13.2")
            }
        }
        val iosMain by getting
        val iosTest by getting
    }
}

android {
    compileSdk = 30
    sourceSets["main"].manifest.srcFile("src/androidMain/AndroidManifest.xml")
    defaultConfig {
        minSdk = 21
        targetSdk = 30
    }
}

```

**Comment by arkivanov** [comment]
First of all, your Compose integration sample has a mistake. If the composable block will be recomposed for any reason, a new instance of the `RootComponent` will be created. Also composition may be executing in a background thread, eventually. It's better to move the component instantiation outside of the composable block.

```kotlin
val component = RootComponent(defaultComponentContext())
setContent {
  NavigationSampleTheme {
    RootUi(component)
  }
}
```

Now regarding the issue with Swift. Seems like in your Gradle file you should add `decompose` module as `api` dependency, and also export it to the Framework. Checkout the following example: https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/common/root/build.gradle.kts

**Reply by falcon4ever** [reply]
Oops, that's a bad novice mistake with Jetpack Compose 😅 thanks for pointing that out!

It looks like the examples are using the "regular framework" for the iOS framework distribution. Based on your suggestion, when I update my gradle file like this, it does work (it also needs `com.arkivanov.essenty:lifecycle` in there):

```kotlin
import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget

plugins {
    kotlin("multiplatform")
    id("com.android.library")
}

kotlin {
    android()

    val iosTarget: (String, KotlinNativeTarget.() -> Unit) -> KotlinNativeTarget = when {
        System.getenv("SDK_NAME")?.startsWith("iphoneos") == true -> ::iosArm64
        else -> ::iosX64
    }

    iosTarget("ios") {
        binaries {
            framework {
                baseName = "shared"
                export("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
                export("com.arkivanov.essenty:lifecycle:0.1.2")
            }
        }
    }

    sourceSets {
        val commonMain by getting {
            dependencies {
                implementation("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
                implementation("com.arkivanov.essenty:lifecycle:0.1.2")

            }
        }
        val commonTest by getting {
            dependencies {
                implementation(kotlin("test-common"))
                implementation(kotlin("test-annotations-common"))
            }
        }
        val androidMain by getting
        val androidTest by getting {
            dependencies {
                implementation(kotlin("test-junit"))
                implementation("junit:junit:4.13.2")
            }
        }
        val iosMain by getting {
            dependencies {
                api("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
                api("com.arkivanov.essenty:lifecycle:0.1.2")
            }
        }
        val iosTest by getting
    }
}

android {
    compileSdk = 30
    sourceSets["main"].manifest.srcFile("src/androidMain/AndroidManifest.xml")
    defaultConfig {
        minSdk = 24
        targetSdk = 30
    }
}
```

My demo projects are using the "CocoaPods dependency manager" since that was the default in the KMM plugin but I unfortunately don't see any option to include the exports in there.

**Reply by arkivanov** [reply]
Unfortunately I'm not familiar with cocoapods, never used this option. And to be honest, I don't see any reason of using a third-party tool if it can be easily avoided. As per [documentation](https://kotlinlang.org/docs/mpp-build-native-binaries.html), I think it should work regardless of the tool you are using. It is just the way how the Framework is packaged.

Just in case, you can omit `lifecycle` and add `transitiveExport = true` instead. Lifecycle is API dependency of Decompose.

**Reply by arkivanov** [reply]
Ahh, seems like exporting is currently available only for "direct integration", for cocoapods there should be a new DSL as well in 1.5.30. Not sure if there are any workarounds currently.

**Reply by falcon4ever** [reply]
I'm not quite sure what the benefit is of using cocoapods to link the shared framework to the iosApp. If you use the "regular framework" it won't prevent you from wrapping the xcodeproj into  a workspace and still use cocoapods (+ update the Podfile manually) so you can still use iOS specific libraries as needed).

On the kotlinlang slack / #multiplatform channel, someone gave me this solution:

If you have this in the "regular framework"
```kotlin
iosTarget("ios") {
    binaries {
        framework {
            baseName = "shared"
            export("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
            export("com.arkivanov.essenty:lifecycle:0.1.2")
        }
    }
}
```

the cocoapods version **hack** for now is:
```kotlin
iosTarget("ios") {
    binaries.getFramework("DEBUG").export("com.arkivanov.decompose:decompose:${rootProject.extra["decomposeVersion"]}")
    binaries.getFramework("DEBUG").export("com.arkivanov.essenty:lifecycle:0.1.2")
}
```

he also told me a similar thing like you mentioned, in kotlin `1.5.30` there will be an option so you can define it as:
```kotlin
cocoapods {
 framework {
  export(...)
 }
}
```

I'll update my kmm demos examples to just use the "regular framework". Seems much simpler and nowadays you dont need the **packforxcode** stuff either.
